File: frontend/iosApp/UILogics/Sources/UILogics/ViewModels/AlbumListViewModel.swift
import Foundation
import Combine
import Domains
import UseCases
import Utilities
@preconcurrency import Shared

@MainActor
public final class AlbumListViewModel: ObservableObject {
    @Published public private(set) var userIcon: UserIconUIModel?
    @Published public private(set) var displayResult: DisplayResult = .loading
    @Published public private(set) var syncState: Shared.SyncQueueState = Shared.SyncQueueState(pendingCount: 0, isSyncing: false)
    @Published public private(set) var isOnline: Bool = true

    public let isNetworkDebugMode: Bool

    private let albumListUseCase: AlbumListUseCaseProtocol
    private let router: AuthenticatedRouterProtocol
    private var cancellables = Set<AnyCancellable>()

    private var isLoadingMore = false

    public init(
        albumListUseCase: AlbumListUseCaseProtocol,
        router: AuthenticatedRouterProtocol,
        isNetworkDebugMode: Bool
    ) {
        self.albumListUseCase = albumListUseCase
        self.router = router
        self.isNetworkDebugMode = isNetworkDebugMode

        albumListUseCase.observeUser()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] user in
                guard let self else { return }
                self.userIcon = UserIconUIModel(
                    avatarUrl: user.displayAvatarURL,
                    didTap: { [weak self] in
                        self?.router.push(.userProfile(user))
                    }
                )
            }
            .store(in: &cancellables)

        albumListUseCase.observeAlbumChange()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] event in
                self?.handleLocalChange(event)
            }
            .store(in: &cancellables)

        albumListUseCase.observeSync()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                self?.syncState = state
            }
            .store(in: &cancellables)

        albumListUseCase.observeOnlineState()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] isOnline in
                self?.isOnline = isOnline
            }
            .store(in: &cancellables)
    }

    public func toggleOnlineState() {
        albumListUseCase.toggleOnlineState()
    }

    private func handleLocalChange(_ event: Shared.LocalAlbumChangeEvent) {
        guard case .success(let currentData) = displayResult else { return }

        var albums = currentData.albums

        switch onEnum(of: event) {
        case .created(let created):
            albums.insert(created.album, at: 0)
        case .updated(let updated):
            if let index = albums.firstIndex(where: { $0.localId == updated.album.localId }) {
                albums[index] = updated.album
            }
        }

        displayResult = .success(makeListData(albums: albums, currentPage: currentData.currentPage, hasMore: currentData.hasMore))
    }

    public func onAppear() {
        guard case .loading = displayResult else { return }
        Task {
            await display()
        }
    }

    public func onRefresh() async {
        await display()
    }

    public func onLoadMore() {
        guard case .success(let listData) = displayResult,
              listData.hasMore,
              !isLoadingMore else { return }
        isLoadingMore = true
        Task {
            await loadMore()
        }
    }

    public func showCreateAlbumForm() {
        router.showSheet(.albumForm(.create))
    }

    public func showEditAlbumForm(album: Album) {
        router.showSheet(.albumForm(.edit(album)))
    }

    public func showSyncQueues() {
        router.push(.syncQueues)
    }

    private func display() async {
        displayResult = .loading
        let result = await albumListUseCase.display()

        switch onEnum(of: result) {
        case .success(let success):
            displayResult = .success(makeListData(albums: success.pageInfo.albums, currentPage: 1, hasMore: success.pageInfo.hasMore))
        case .failure(let failure):
            displayResult = .failure(mapDisplayError(failure.error))
        }
    }

    private func loadMore() async {
        guard case .success(let currentData) = displayResult else { return }

        let previousCount = currentData.albums.count
        let nextPage = currentData.currentPage + 1
        let result = await albumListUseCase.next(page: nextPage)

        switch onEnum(of: result) {
        case .success(let success):
            displayResult = .success(makeListData(albums: success.pageInfo.albums, currentPage: nextPage, hasMore: success.pageInfo.hasMore))

            // If data didn't increase but hasMore is true, fetch next page automatically
            if success.pageInfo.albums.count == previousCount && success.pageInfo.hasMore {
                await loadMore()
            } else {
                isLoadingMore = false
            }
        case .failure:
            isLoadingMore = false
        }
    }

    private func makeListData(albums: [Album], currentPage: Int, hasMore: Bool) -> ListData {
        let items = albums.map { album in
            AlbumItemUIModel(
                id: album.localIdUUID,
                title: album.title,
                coverImageUrl: album.displayCoverImageURL,
                syncStatus: album.syncStatus,
                didTap: { [weak self] in
                    self?.router.push(.albumDetail(.albumList(album)))
                }
            )
        }
        return ListData(albums: albums, items: items, currentPage: currentPage, hasMore: hasMore)
    }

    private func mapDisplayError(_ error: Shared.AlbumDisplayError) -> ErrorUIModel {
        switch error {
        case .networkError:
            return ErrorUIModel(
                message: "Network error. Please check your connection.",
                retryAction: { [weak self] in
                    Task { await self?.display() }
                }
            )
        case .offline:
            return ErrorUIModel(
                message: "You're offline and have no cached albums.",
                retryAction: { [weak self] in
                    Task { await self?.display() }
                }
            )
        default:
            return ErrorUIModel(
                message: "An unexpected error occurred.",
                retryAction: { [weak self] in
                    Task { await self?.display() }
                }
            )
        }
    }
}

// MARK: - UI Models

extension AlbumListViewModel {
    public struct UserIconUIModel: Equatable, Sendable {
        public let avatarUrl: URL?
        @EquatableNoop public var didTap: @MainActor @Sendable () -> Void

        public init(avatarUrl: URL?, didTap: @escaping @MainActor @Sendable () -> Void) {
            self.avatarUrl = avatarUrl
            self.didTap = didTap
        }
    }

    public enum DisplayResult: Equatable {
        case loading
        case success(ListData)
        case failure(ErrorUIModel)
    }

    public struct ListData: Equatable {
        public let albums: [Album]
        public let items: [AlbumItemUIModel]
        public let currentPage: Int
        public let hasMore: Bool

        public init(albums: [Album], items: [AlbumItemUIModel], currentPage: Int, hasMore: Bool) {
            self.albums = albums
            self.items = items
            self.currentPage = currentPage
            self.hasMore = hasMore
        }
    }

    public struct AlbumItemUIModel: Equatable, Identifiable {
        public let id: UUID
        public let title: String
        public let coverImageUrl: URL?
        public let syncStatus: SyncStatus
        @EquatableNoop public var didTap: @MainActor () -> Void

        public init(id: UUID, title: String, coverImageUrl: URL?, syncStatus: SyncStatus, didTap: @escaping @MainActor () -> Void) {
            self.id = id
            self.title = title
            self.coverImageUrl = coverImageUrl
            self.syncStatus = syncStatus
            self.didTap = didTap
        }
    }

    public struct ErrorUIModel: Equatable {
        public let message: String
        @EquatableNoop public var retryAction: @MainActor () -> Void

        public init(message: String, retryAction: @escaping @MainActor () -> Void) {
            self.message = message
            self.retryAction = retryAction
        }
    }
}

File: frontend/composeApp/src/androidMain/kotlin/com/example/memoriesapp/ui/uilogics/viewmodels/AlbumListViewModel.kt
package com.example.memoriesapp.ui.uilogics.viewmodels

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.memoriesapp.core.SyncStatus
import com.example.memoriesapp.domain.Album
import com.example.memoriesapp.repository.LocalAlbumChangeEvent
import com.example.memoriesapp.repository.SyncQueueState
import com.example.memoriesapp.usecase.AlbumDisplayError
import com.example.memoriesapp.usecase.AlbumDisplayResult
import com.example.memoriesapp.usecase.AlbumListUseCaseWrapper
import com.example.memoriesapp.usecase.AlbumNextResult
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

/**
 * ViewModel for album list screen.
 */
class AlbumListViewModel(
    private val albumListUseCase: AlbumListUseCaseWrapper
) : ViewModel() {

    var userAvatarUrl by mutableStateOf<String?>(null)
        private set
    var displayResult by mutableStateOf<DisplayResult>(DisplayResult.Loading)
        private set
    var syncState by mutableStateOf(SyncQueueState(pendingCount = 0, isSyncing = false))
        private set
    var isOnline by mutableStateOf(true)
        private set

    val isNetworkDebugMode: Boolean = albumListUseCase.isNetworkDebugMode

    // Pagination state
    var isLoadingMore by mutableStateOf(false)
        private set

    private val _navigationEvent = MutableSharedFlow<NavigationEvent>()
    val navigationEvent: SharedFlow<NavigationEvent> = _navigationEvent.asSharedFlow()

    init {
        observeUser()
        observeAlbumChange()
        observeSync()
        observeOnlineState()
    }

    private fun observeUser() {
        albumListUseCase.observeUser()
            .onEach { user ->
                userAvatarUrl = user.displayAvatar
            }
            .launchIn(viewModelScope)
    }

    private fun observeAlbumChange() {
        albumListUseCase.observeAlbumChange()
            .onEach { event ->
                handleLocalChange(event)
            }
            .launchIn(viewModelScope)
    }

    private fun observeSync() {
        albumListUseCase.observeSync()
            .onEach { state ->
                syncState = state
            }
            .launchIn(viewModelScope)
    }

    private fun observeOnlineState() {
        albumListUseCase.observeOnlineState()
            .onEach { online ->
                isOnline = online
            }
            .launchIn(viewModelScope)
    }

    fun toggleOnlineState() {
        albumListUseCase.toggleOnlineState()
    }

    private fun handleLocalChange(event: LocalAlbumChangeEvent) {
        val currentData = (displayResult as? DisplayResult.Success)?.data ?: return
        val albums = currentData.albums.toMutableList()

        when (event) {
            is LocalAlbumChangeEvent.Created -> {
                albums.add(0, event.album)
            }
            is LocalAlbumChangeEvent.Updated -> {
                val index = albums.indexOfFirst { it.localId == event.album.localId }
                if (index >= 0) {
                    albums[index] = event.album
                }
            }
        }

        displayResult = DisplayResult.Success(
            makeListData(albums, currentData.currentPage, currentData.hasMore)
        )
    }

    fun onAppear() {
        if (displayResult is DisplayResult.Loading) {
            viewModelScope.launch {
                display()
            }
        }
    }

    fun onRefresh() {
        viewModelScope.launch {
            display()
        }
    }

    fun onLoadMore() {
        val currentData = (displayResult as? DisplayResult.Success)?.data ?: return
        if (!currentData.hasMore || isLoadingMore) return

        viewModelScope.launch {
            loadMore()
        }
    }

    fun onAlbumTap(album: Album) {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.AlbumDetail(album))
        }
    }

    fun showCreateAlbumForm() {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.CreateAlbum)
        }
    }

    fun showEditAlbumForm(album: Album) {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.EditAlbum(album))
        }
    }

    fun showUserProfile() {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.UserProfile)
        }
    }

    fun showSyncQueues() {
        viewModelScope.launch {
            _navigationEvent.emit(NavigationEvent.SyncQueues)
        }
    }

    private suspend fun display() {
        displayResult = DisplayResult.Loading
        when (val result = albumListUseCase.display()) {
            is AlbumDisplayResult.Success -> {
                displayResult = DisplayResult.Success(
                    makeListData(result.pageInfo.albums, 1, result.pageInfo.hasMore)
                )
            }
            is AlbumDisplayResult.Failure -> {
                displayResult = DisplayResult.Failure(mapDisplayError(result.error))
            }
        }
    }

    private suspend fun loadMore() {
        val currentData = (displayResult as? DisplayResult.Success)?.data ?: return
        val previousCount = currentData.albums.size
        val nextPage = currentData.currentPage + 1

        isLoadingMore = true

        when (val result = albumListUseCase.next(nextPage)) {
            is AlbumNextResult.Success -> {
                displayResult = DisplayResult.Success(
                    makeListData(result.pageInfo.albums, nextPage, result.pageInfo.hasMore)
                )
                // If data didn't increase but hasMore is true, fetch next page automatically
                if (result.pageInfo.albums.size == previousCount && result.pageInfo.hasMore) {
                    loadMore()
                } else {
                    isLoadingMore = false
                }
            }
            is AlbumNextResult.Failure -> {
                isLoadingMore = false
            }
        }
    }

    private fun makeListData(albums: List<Album>, currentPage: Int, hasMore: Boolean): ListData {
        val items = albums.map { album ->
            AlbumItemUIModel(
                localId = album.localId.toString(),
                title = album.title,
                coverImageUrl = album.displayCoverImage,
                syncStatus = album.syncStatus
            )
        }
        return ListData(albums, items, currentPage, hasMore)
    }

    private fun mapDisplayError(error: AlbumDisplayError): ErrorUIModel {
        return when (error) {
            AlbumDisplayError.NETWORK_ERROR -> ErrorUIModel(
                message = "Network error. Please check your connection.",
                onRetry = { viewModelScope.launch { display() } }
            )
            AlbumDisplayError.OFFLINE -> ErrorUIModel(
                message = "You're offline and have no cached albums.",
                onRetry = { viewModelScope.launch { display() } }
            )
            AlbumDisplayError.UNKNOWN -> ErrorUIModel(
                message = "An unexpected error occurred.",
                onRetry = { viewModelScope.launch { display() } }
            )
        }
    }

    sealed class DisplayResult {
        data object Loading : DisplayResult()
        data class Success(val data: ListData) : DisplayResult()
        data class Failure(val error: ErrorUIModel) : DisplayResult()
    }

    data class ListData(
        val albums: List<Album>,
        val items: List<AlbumItemUIModel>,
        val currentPage: Int,
        val hasMore: Boolean
    )

    data class AlbumItemUIModel(
        val localId: String,
        val title: String,
        val coverImageUrl: String?,
        val syncStatus: SyncStatus
    )

    data class ErrorUIModel(
        val message: String,
        val onRetry: () -> Unit
    )

    sealed class NavigationEvent {
        data class AlbumDetail(val album: Album) : NavigationEvent()
        data object CreateAlbum : NavigationEvent()
        data class EditAlbum(val album: Album) : NavigationEvent()
        data object UserProfile : NavigationEvent()
        data object SyncQueues : NavigationEvent()
    }
}
